filebeat/reader/readfile/scanner.go

```diff
@@ -0,0 +1,110 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package readfile
+
+import (
+	"io"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+// lineScanner scans for new lines separated by a configurable byte slice.
+type lineScanner struct {
+	in         *decoderReader
+	separator  []byte
+	bufferSize int
+
+	buf           *streambuf.Buffer
+	segmentOffset int
+	streamOffset  int
+}
+
+func newLineScanner(in *decoderReader, separator []byte, bufferSize int) *lineScanner {
+	return &lineScanner{
+		in:         in,
+		separator:  separator,
+		bufferSize: bufferSize,
+		buf:        streambuf.New(nil),
+	}
+}
+
+// seekToLastRead moves the fp to the last read position.
+// The number of bytes which needs to be read is the size of the
+// configured harvester_buffer_size when the state was written.
+func (s *lineScanner) seekToLastRead() error {
```

Comment:
why do we need to seek in the line scanner? Seeking should be done in the most lowest input layer of possible.

Reply:
To me it seems logically cleaner, because `lineScanner` tracks the `Converted*Offset`s. But it's true it's better to seek in the decoder, so a function calls back to `decoderReader` could be saved.